src/sasl/cyrsasl_external.erl

```diff
@@ -44,33 +45,69 @@ stop() ->
 
 -spec mech_new(Host :: ejabberd:server(),
                Creds :: mongoose_credentials:t()) -> {ok, sasl_external_state()}.
-mech_new(_Host, Creds) ->
+mech_new(Host, Creds) ->
     Cert = mongoose_credentials:get(Creds, client_cert, no_cert),
     maybe_extract_certs(Cert, Creds).
 
 maybe_extract_certs(no_cert, Creds) ->
-    {ok, #state{creds = Creds}};
+    {ok, #state{creds = mongoose_credentials:extend(Creds, [{cert_file, false}])}};
 maybe_extract_certs(Cert, Creds) ->
     DerCert = public_key:pkix_encode('Certificate', Cert, plain),
     PemCert = public_key:pem_encode([{'Certificate', DerCert, not_encrypted}]),
     CertFields = get_common_name(Cert) ++ get_xmpp_addresses(Cert),
-    SaslExternalCredentials = [{pem_cert, PemCert}, {der_cert, DerCert} | CertFields],
+    SaslExternalCredentials = [{cert_file, true}, {pem_cert, PemCert}, {der_cert, DerCert} | CertFields],
     {ok, #state{creds = mongoose_credentials:extend(Creds, SaslExternalCredentials)}}.
 
 
 -spec mech_step(State :: sasl_external_state(),
                 ClientIn :: binary()) -> {ok, mongoose_credentials:t()} | {error, binary()}.
-mech_step(#state{creds = Creds}, <<"">>) ->
-    authorize(Creds);
 mech_step(#state{creds = Creds}, User) ->
-    authorize(mongoose_credentials:set(Creds, requested_name, User)).
-
-authorize(Creds) ->
-    %% auth backend is responsible to add username to Creds.
-    case ejabberd_auth:authorize(Creds) of
-        {ok, NewCreds} -> {ok, NewCreds};
-        {error, invalid_authid} -> {error, <<"invalid-authzid">>};
-        _ -> {error, <<"not-authorized">>}
+    case mongoose_credentials:get(Creds, cert_file) of
+        false ->
+            {error, <<"not-authorized">>};
+        true ->
+            do_mech_step(Creds, User)
+    end.
+do_mech_step(Creds, User) ->
+    XmppAddrs = get_credentials(Creds, xmpp_addresses),
+    CommonName = get_credentials(Creds, common_name),
+    Server = mongoose_credentials:lserver(Creds),
+    case check_auth_req(XmppAddrs, CommonName, User, Server) of
+        {error, Error} ->
+            {error, Error};
+        {ok, Name} ->
+            NewCreds = mongoose_credentials:extend(Creds, [{username, Name}]),
+            ejabberd_auth:authorize(NewCreds)
```

Comment:
This function may return `{error, not_authorized}`, why the specs in line 63 say `{error, binary()}`.